raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java

```diff
@@ -356,6 +351,299 @@ public void testEndQuorumStartsNewElectionImmediatelyIfFollowerUnattached() thro
         assertEquals(ElectionState.withVotedCandidate(epoch + 1, localId, voters), quorumStateStore.readElectionState());
     }
 
+    @Test
+    public void testLocalReadFromLeader() throws Exception {
+        int otherNodeId = 1;
+        int epoch = 2;
+
+        Set<Integer> voters = Utils.mkSet(localId, otherNodeId);
+        KafkaRaftClient client = initializeAsLeader(voters, epoch);
+        assertEquals(1L, log.endOffset().offset);
+        assertEquals(OptionalLong.empty(), client.highWatermark());
+
+        deliverRequest(fetchRequest(epoch, otherNodeId, 1L, epoch, 0));
+        client.poll();
+        assertEquals(1L, log.endOffset().offset);
+        assertEquals(OptionalLong.of(1L), client.highWatermark());
+        assertSentFetchResponse(Errors.NONE, epoch, OptionalInt.of(localId));
+
+        SimpleRecord[] records = new SimpleRecord[] {
+            new SimpleRecord("a".getBytes()),
+            new SimpleRecord("b".getBytes())
+        };
+        client.append(MemoryRecords.withRecords(CompressionType.NONE, records), AckMode.LEADER, Integer.MAX_VALUE);
+        client.poll();
+        assertEquals(3L, log.endOffset().offset);
+        assertEquals(OptionalLong.of(1L), client.highWatermark());
+
+        validateLocalRead(client, new OffsetAndEpoch(1L, epoch), Isolation.COMMITTED, new SimpleRecord[0]);
+        validateLocalRead(client, new OffsetAndEpoch(1L, epoch), Isolation.UNCOMMITTED, records);
+        validateLocalRead(client, new OffsetAndEpoch(3L, epoch), Isolation.COMMITTED, new SimpleRecord[0]);
+        validateLocalRead(client, new OffsetAndEpoch(3L, epoch), Isolation.UNCOMMITTED, new SimpleRecord[0]);
+
+        deliverRequest(fetchRequest(epoch, otherNodeId, 3L, epoch, 0));
+        client.poll();
+        assertEquals(OptionalLong.of(3L), client.highWatermark());
+        assertSentFetchResponse(Errors.NONE, epoch, OptionalInt.of(localId));
+
+        validateLocalRead(client, new OffsetAndEpoch(1L, epoch), Isolation.COMMITTED, records);
+    }
+
+    @Test
+    public void testDelayedLocalReadFromLeader() throws Exception {
+        int otherNodeId = 1;
+        int epoch = 2;
+
+        Set<Integer> voters = Utils.mkSet(localId, otherNodeId);
+        KafkaRaftClient client = initializeAsLeader(voters, epoch);
+        assertEquals(1L, log.endOffset().offset);
+        assertEquals(OptionalLong.empty(), client.highWatermark());
+
+        deliverRequest(fetchRequest(epoch, otherNodeId, 1L, epoch, 0));
+        client.poll();
+        assertEquals(1L, log.endOffset().offset);
+        assertEquals(OptionalLong.of(1L), client.highWatermark());
+        assertSentFetchResponse(Errors.NONE, epoch, OptionalInt.of(localId));
+
+        CompletableFuture<Records> logEndReadFuture = client.read(new OffsetAndEpoch(1L, epoch),
```

Comment:
nit: maybe we can wrap those `assertFalse` cases into a new private function similar to `validateLocalRead` as well, like `validateLocalReadFailed`.